packages/wonder-blocks-button/components/clickable-behavior.js

```diff
@@ -0,0 +1,190 @@
+ */
+import React from "react";
+
+type Props = {
+    children: (state: State, handlers: Handlers) => React$Element<*>,
+    disabled: boolean,
+    href?: string,
+    onClick?: (e: SyntheticMouseEvent<>) => void,
+};
+
+type State = {
+    hovered: boolean,
+    focused: boolean,
+    pressed: boolean,
+};
+
+type Handlers = {
+    onClick: (e: SyntheticMouseEvent<>) => void,
+    onMouseEnter: () => void,
+    onMouseLeave: () => void,
+    onMouseDown: () => void,
+    onMouseUp: () => void,
+    onTouchStart: () => void,
+    onTouchEnd: () => void,
+    onKeyDown: (e: KeyboardEvent) => void,
+    onKeyUp: (e: KeyboardEvent) => void,
+    onBlur: () => void,
+};
+
+const keyCodes = {
+    tab: 9,
+    enter: 13,
+    space: 32,
+};
+
+export default class ClickableBehavior extends React.Component<Props, State> {
+    constructor(props: Props) {
+        super(props);
+
+        this.state = {
+            hovered: false,
+            focused: false,
+            pressed: false,
+        };
+    }
+
+    waitingForClick: boolean;
+
+    handleClick = (e: SyntheticMouseEvent<>) => {
+        if (!this.props.disabled && this.props.onClick) {
+            this.waitingForClick = false;
+            this.props.onClick(e);
+        }
+    };
+
+    handleMouseEnter = () => {
+        if (!this.props.disabled && !this.waitingForClick) {
+            this.setState({hovered: true});
+        }
+    };
+
+    handleMouseLeave = () => {
+        if (!this.props.disabled && !this.waitingForClick) {
+            // TODO(yejia): Is this what we want for pressed?
+            this.setState({hovered: false, pressed: false});
+        }
+    };
+
+    handleMouseDown = () => {
+        if (!this.props.disabled) {
+            this.setState({pressed: true});
+        }
+    };
+
+    handleMouseUp = () => {
+        if (!this.props.disabled) {
+            this.setState({pressed: false});
+        }
+    };
+
+    handleTouchStart = () => {
+        if (!this.props.disabled) {
+            this.setState({pressed: true});
+        }
+    };
+
+    handleTouchEnd = () => {
+        if (!this.props.disabled) {
+            this.setState({pressed: false});
+            this.waitingForClick = true;
+        }
+    };
+
+    handleKeyDown = (event: KeyboardEvent) => {
+        const keyCode = event.which || event.keyCode;
+        const {disabled, href} = this.props;
+        if (!disabled) {
+            if (keyCode === keyCodes.tab) {
+                this.setState({focused: false});
+            } else if (
+                href ? keyCode === keyCodes.enter : keyCode === keyCodes.space
+            ) {
+                this.setState({pressed: true});
+            }
+        }
+    };
+
+    handleKeyUp = (event: KeyboardEvent) => {
+        const keyCode = event.which || event.keyCode;
+        const {disabled, href} = this.props;
+        if (!disabled) {
+            if (keyCode === keyCodes.tab) {
+                this.setState({focused: true});
```

Comment:
I wasn't able to get the buttons to focus with the "tab" key in Safari.  It works like a charm in Chrome though.

Reply:
I wonder if this is even necessary since tabbing should focus a button by default.  All we should need to do is track if the button has become focused because of a keyboard event.

Reply:
Hmmm I don't know how to do that :( I don't think I can just check the type of event passed in to onFocus. In general I don't know how to check type of SyntheticEvents, but also I think SyntheticFocusEvents aren't SyntheticKeyboardEvents. Also tabbing does work for me in Safari, so not sure how to check what's going on x_x

Reply:
It looks like it isn't turned on by default in Safari.  After turning it on, it works as expected.  I'm not sure why they'd turn it off by default.  :(

I was going to suggest tracking if there was a `mousedown` or `touchstart` event on the button and if there wasn't then we could assume the focus was caused by the keyboard.  Totally unnecessary since the problem was a Safari setting.